api/python/t4/packages.py

```diff
@@ -436,53 +434,49 @@ def set(self, logical_key, entry=None, meta=None):
                 else in the entry
 
         Returns:
-            A new package
+            self
         """
         if entry is None and meta is None:
             raise PackageException('Must specify either entry or meta')
 
         if entry is None:
             return self._update_meta(logical_key, meta)
 
-        pkg = self._clone()
         if isinstance(entry, str):
             entry = PackageEntry.from_local_path(entry)
             if meta is not None:
                 entry.meta = meta
-            pkg._data[logical_key] = entry
+            self._data[logical_key] = entry
         elif isinstance(entry, PackageEntry):
-            pkg._data[logical_key] = entry
             if meta is not None:
                 raise PackageException("Must specify metadata in the entry")
+            self._data[logical_key] = entry
         else:
             raise NotImplementedError
-        
-        # modified package should have new top hash
-        pkg._unset_tophash()
-        return pkg
+
+        # must unset old top hash when modifying package
+        self._unset_tophash()
+        return self
 
     def _update_meta(self, logical_key, meta):
-        pkg = self._clone()
-        pkg._data[logical_key].meta = meta
-        # modified package should have new top hash
-        pkg._unset_tophash()
-        return pkg
+        self._data[logical_key].meta = meta
+        self._unset_tophash()
+        return self
 
     def delete(self, logical_key):
         """
-        Returns a new package with logical_key removed.
+        Returns the package with logical_key removed.
 
         Returns:
-            A new package
+            self
 
         Raises:
             KeyError: when logical_key is not present to be deleted
         """
-        pkg = self._clone()
-        pkg._data.pop(logical_key)
-        # modified package should have new top hash
-        pkg._unset_tophash()
-        return pkg
+        self._data.pop(logical_key)
+        # must unset old top hash when modifying package
+        self._unset_tophash()
```

Comment:
Can we add a test that the tophash changes when we do these sensitive operations (delete, set, update, capture)